src/hotspot/share/runtime/flags/jvmFlagLimit.cpp

```diff
@@ -0,0 +1,162 @@
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "memory/allocation.inline.hpp"
+#include "gc/shared/jvmFlagConstraintsGC.hpp"
+#include "runtime/flags/jvmFlag.hpp"
+#include "runtime/flags/jvmFlagLimit.hpp"
+#include "runtime/flags/jvmFlagConstraintList.hpp"
+#include "runtime/flags/jvmFlagConstraintsCompiler.hpp"
+#include "runtime/flags/jvmFlagConstraintsRuntime.hpp"
+#include "runtime/flags/jvmFlagRangeList.hpp"
+#include "runtime/globals_extension.hpp"
+#include "gc/shared/referenceProcessor.hpp"
+#include "oops/markWord.hpp"
+#include "runtime/task.hpp"
+
+//----------------------------------------------------------------------
+// Build flagLimitTable[]
+
+#define CONSTRAINT_ENUM(func)         constraint_enum_ ## func
+#define CONSTRAINT_ENUM_(type, func)  CONSTRAINT_ENUM(func),
+#define CONSTRAINT_FUNC(type, func)   (void*)&func,
+
+enum JVMFlagConstraintsEnum {
+  ALL_CONSTRAINTS(CONSTRAINT_ENUM_)
+  NUM_JVMFlagConstraintsEnum
+};
+
+static void* const flagConstraintTable[NUM_JVMFlagConstraintsEnum] = {
+  ALL_CONSTRAINTS(CONSTRAINT_FUNC)
+};
+
+void* JVMFlagLimit::constraint_func() const {
+  int i = _constraint_func;
+  assert(0 <= i && i < NUM_JVMFlagConstraintsEnum, "sanity");
+  return flagConstraintTable[i];
+}
+
+struct DummyLimit {
+  char dummy;
+  constexpr DummyLimit(...) : dummy() {}
+};
+
+template <typename T>
+class LimitGetter {
+public:
+  // These functions return NULL for develop flags in a PRODUCT build
+  static constexpr const JVMFlagLimit* no_limit(int dummy) {
+    return NULL;
+  }
+  static constexpr const JVMFlagLimit* no_limit(int dummy, T min, T max) {
+    return NULL;
+  }
+  static constexpr const JVMFlagLimit* no_limit(int dummy, ConstraintMarker dummy2, short func, int phase) {
+    return NULL;
+  }
+  static constexpr const JVMFlagLimit* no_limit(int dummy, T min, T max, ConstraintMarker dummy2, short func, int phase) {
+    return NULL;
+  }
+  static constexpr const JVMFlagLimit* no_limit(int dummy, ConstraintMarker dummy2, short func, int phase, T min, T max) {
+    return NULL;
+  }
+
+  // This is for flags that have neither range no constraint. We don't need the JVMFlagLimit struct.
+  static constexpr const JVMFlagLimit* get_limit(const JVMTypedFlagLimit<T>* p, int dummy) {
+    return NULL;
+  }
+
+  static constexpr const JVMFlagLimit* get_limit(const JVMTypedFlagLimit<T>* p, int dummy, T min, T max) {
+    return p;
+  }
+  static constexpr const JVMFlagLimit* get_limit(const JVMTypedFlagLimit<T>* p, int dummy, ConstraintMarker dummy2, short func, int phase) {
+    return p;
+  }
+  static constexpr const JVMFlagLimit* get_limit(const JVMTypedFlagLimit<T>* p, int dummy, T min, T max, ConstraintMarker dummy2, short func, int phase) {
+    return p;
+  }
+  static constexpr const JVMFlagLimit* get_limit(const JVMTypedFlagLimit<T>* p, int dummy, ConstraintMarker dummy2, short func, int phase, T min, T max) {
+    return p;
+  }
```

Comment:
Instead of the 'no_limit' functions, could you have get_limit have PRODUCT_RETURN(NULL) ?  Otherwise, fine to get the macros working.

Reply:
I think your suggestion would simply the macro materialization from
```
#ifdef PRODUCT
  ALL_FLAGS(FLAG_LIMIT_PTR_NONE,
            FLAG_LIMIT_PTR_NONE,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR_NONE,
            APPLY_FLAG_RANGE,
            APPLY_FLAG_CONSTRAINT)
#else
  ALL_FLAGS(FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            APPLY_FLAG_RANGE,
            APPLY_FLAG_CONSTRAINT)
#endif
```
to 
```
ALL_FLAGS(FLAG_LIMIT_PTR_NOT_PRODUCT,
            FLAG_LIMIT_PTR_NOT_PRODUCT,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR,
            FLAG_LIMIT_PTR_NOT_PRODUCT,
            APPLY_FLAG_RANGE,
            APPLY_FLAG_CONSTRAINT)
```
But I would need twice the number of the get_limit functions, like:
```
  static constexpr const JVMFlagLimit* get_limit(const JVMTypedFlagLimit<T>* p, int dummy, T min, T max) {
    return p;    // return product flag for all builds
  }
  static constexpr const JVMFlagLimit* get_limit_not_product(const JVMTypedFlagLimit<T>* p, int dummy, T min, T max) {
    NOT_PRODUCT(return p);     // return non-product flag only in non-product builds
    PRODUCT_ONLY(return NULL);  // .... but not in product build
  }
```
So the code will be more verbose than before.

Instead, in [bf90a37](https://github.com/openjdk/jdk/pull/82/commits/bf90a3765720267a8770f7bb0cc6957471ccb21b), I changed no_limit to a varargs function so now there's only one version of it.